### PR TITLE
Added app for EVP_KDF

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man1/kdf.pod [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+=head1 NAME
+
+openssl-kdf,
+kdf - perform Key Derivation Function operations
+
+=head1 SYNOPSIS
+
+B<openssl kdf>
+[B<-help>]
+[B<-kdfopt>]
```
Comment (levitte) [inline doc/man1/kdf.pod]: Make that `[B<-kdfopt> I<nm:v>]`

File: doc/man1/kdf.pod [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+[B<-keylen num>]
+[B<-out filename>]
+[B<-binary>]
+B<kdf_name>
```
Comment (levitte) [inline doc/man1/kdf.pod]: According to both Linux' [man-pages(7)](http://man7.org/linux/man-pages/man7/man-pages.7.html) and FreeBSD's [man(7)](https://www.freebsd.org/cgi/man.cgi?query=man&sektion=7), arguments should be in italics, so the above four lines should really be:

```
[B<-keylen> I<num>]
[B<-out> I<filename>]
[B<-binary>]
I<kdf_name>
```

I know that we haven't done this in most of our pages, but perhaps we should?

File: doc/man1/kdf.pod [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+B<openssl> I<kdf> [B<...>] B<kdf_name>
```
Comment (levitte) [inline doc/man1/kdf.pod]: Why this line at all?

File: doc/man1/kdf.pod [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+Specifies the name of a supported KDF algorithm which will be used.
+The supported algorithms are TLS1-PRF, HKDF, SSKDF, PBKDF2 and id-scrypt.
+
+=back
+
+=head1 EXAMPLES
+
+Use TLS1-PRF to create a hex-encoded derived key from a secret key and seed: \
+openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt key:secret \
+        -kdfopt seed:seed TLS1-PRF
```
Comment (levitte) [inline doc/man1/kdf.pod]: Not sure I understand what you want to do here...  I think the section should be formatted like this:

```
Use TLS1-PRF to create a hex-encoded derived key from a secret key and seed:

    openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt key:secret \
            -kdfopt seed:seed TLS1-PRF
```

Comment (levitte) [reply doc/man1/kdf.pod]: ... and similarly for all other examples

**Comment by levitte** [comment]
I see your approach with saving all option data first, and only act on them after all arguments have been processed.  This is a huge contrast to, for example, `openssl genpkey`, which acts immediately on the options (which means the options must come in a specific order).

We should probably think about streamlining all apps to behave the same way...

**Comment by simo5** [comment]
Any reason to leave out SSHKDF from test and docs ?

**Comment by slontis** [comment]
I forgot about SSHKDF (I think mainly because the name it looks a lot like SSKDF). Will add.

**Comment by slontis** [comment]
It appears there is an inconsistency with the kdf docs related to md name. I think David changed it to use 'digest' instead of 'md' (but didnt update the docs). 
@simo5 I dont think this PR is the place to do this, but could you fix this up please? There is also a evp_kdf_test.c that SSHKDF should be added to.

File: doc/man1/kdf.pod (hunk 4 above)
Comment (slontis) [reply doc/man1/kdf.pod]: Hmm still looks bad.. I changed it to this.

=head2 Use TLS1-PRF to create a hex-encoded derived key from a secret key and seed:
C<openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt key:secret -kdfopt seed:seed TLS1-PRF>

File: doc/man1/kdf.pod (hunk 3 above)
Comment (slontis) [reply doc/man1/kdf.pod]: removed

File: test/recipes/20-test_kdf.t [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+use OpenSSL::Test;
+use OpenSSL::Test::Utils;
+
+setup("test_kdf");
+
+my @kdf_tests = (
+    { cmd => [qw{openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt secret:secret -kdfopt seed:seed TLS1-PRF}],
+      expected => '8E:4D:93:25:30:D7:65:A0:AA:E9:74:C3:04:73:5E:CC',
+      desc => 'TLS1-PRF SHA256' },
```
Comment (tomato42) [inline test/recipes/20-test_kdf.t]: maybe also add test with TLS 1.0/TLS 1.1 PRF, as they use the combined MD5-SHA1 hash – I'm not sure if it doesn't require special handling

File: doc/man1/kdf.pod [hunk 6]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the name of a digest as an alphanumeric string.
+To see the list of supported digests, use the command I<list -digest-commands>.
+
+=back
+
+=item B<kdf_name>
+
+Specifies the name of a supported KDF algorithm which will be used.
+The supported algorithms are TLS1-PRF, HKDF, SSKDF, PBKDF2, SSHKDF and id-scrypt.
```
Comment (tomato42) [inline doc/man1/kdf.pod]: why the inconsistency for referring to scrypt?

**Comment by simo5** [comment]
> It appears there is an inconsistency with the kdf docs related to md name. I think David changed it to use 'digest' instead of 'md' (but didnt update the docs).

Is this referred to sshkdf specifically? If so where do I need to change things?

> @simo5 I dont think this PR is the place to do this, but could you fix this up please? There is also a evp_kdf_test.c that SSHKDF should be added to.

Will add a PR shortly that handles this.

**Comment by simo5** [comment]
Added evp_kdf_test in #8774

File: doc/man1/kdf.pod [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+
+openssl-kdf,
+kdf - perform Key Derivation Function operations
+
+=head1 SYNOPSIS
+
+B<openssl kdf>
+[B<-help>]
+[B<-kdfopt> I<nm:v>]
+[B<-keylen I<num>>]
```
Comment (levitte) [inline doc/man1/kdf.pod]: Syntax error

File: doc/man1/kdf.pod [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+[B<-out I<filename>>]
```
Comment (levitte) [inline doc/man1/kdf.pod]: Here as well

File: doc/man1/kdf.pod [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+
+=head1 OPTIONS
+
+=over 4
+
+=item B<-help>
+
+Print a usage message.
+
+=item B<-keylen num>
```
Comment (levitte) [inline doc/man1/kdf.pod]: `B<-keylen> I<num>`

File: doc/man1/kdf.pod [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+
+The output size of the derived key. This field is required.
+
+=item B<-out filename>
```
Comment (levitte) [inline doc/man1/kdf.pod]: Similarly here

File: doc/man1/kdf.pod [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+
+Filename to output to, or standard output by default.
+
+=item B<-binary>
+
+Output the derived key in binary form. Uses hexadecimal text format if not specified.
+
+=item B<-kdfopt nm:v>
```
Comment (levitte) [inline doc/man1/kdf.pod]: ... and here

File: doc/man1/kdf.pod [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+
+Passes options to the KDF algorithm.
+A comprehensive list of controls can be found in the EVP_KDF_CTX implementation
+documentation.
+Common control strings used by EVP_KDF_ctrl_str() are:
+
+=over 4
+
+=item B<key:string>
```
Comment (levitte) [inline doc/man1/kdf.pod]: If we want to be strict about it, this should be `B<key:>I<string>`

File: doc/man1/kdf.pod [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the secret key as an alphanumeric string (use if the key contains
+printable characters only).
+The string length must conform to any restrictions of the KDF algorithm.
+A key must be specified for most KDF algorithms.
+
+=item B<hexkey:string>
```
Comment (levitte) [inline doc/man1/kdf.pod]: Likewise here

File: doc/man1/kdf.pod [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the secret key in hexadecimal form (two hex digits per byte).
+The key length must conform to any restrictions of the KDF algorithm.
+A key must be specified for most KDF algorithms.
+
+=item B<pass:string>
```
Comment (levitte) [inline doc/man1/kdf.pod]: ... and here

File: doc/man1/kdf.pod [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the password as an alphanumeric string (use if the password contains
+printable characters only).
+The password must be specified for PBKDF2 and scrypt.
+
+=item B<hexpass:string>
```
Comment (levitte) [inline doc/man1/kdf.pod]: ... and here

File: doc/man1/kdf.pod [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the password in hexadecimal form (two hex digits per byte).
+The password must be specified for PBKDF2 and scrypt.
+
+=item B<digest:string>
```
Comment (levitte) [inline doc/man1/kdf.pod]: ... and here

File: doc/man1/kdf.pod [hunk 17]
```diff
@@ -0,0 +1,152 @@
...
+
+Specifies the name of a digest as an alphanumeric string.
+To see the list of supported digests, use the command I<list -digest-commands>.
+
+=back
+
+=item B<kdf_name>
```
Comment (levitte) [inline doc/man1/kdf.pod]: `I<kdf_name>`

File: doc/man1/kdf.pod [hunk 18]
```diff
@@ -0,0 +1,152 @@
...
+C<openssl kdf -keylen 16 -kdfopt md:SHA256 -kdfopt hexkey:0102030405 -kdfopt hexxcghash:06090A -kdfopt hexsession_id:01020304 -kdfopt type:A SSHKDF>
+
+=head2 Use PBKDF2 to create a hex-encoded derived key from a password and salt:
+
+C<openssl kdf -keylen 32 -kdfopt digest:SHA256 -kdfopt pass:password -kdfopt salt:salt -kdfopt iter:2 PBKDF2>
+
+=head2 Use scrypt to create a hex-encoded derived key from a password and salt:
+
+C<openssl kdf -keylen 64 -kdfopt pass:password -kdfopt salt:NaCl -kdfopt N:1024 -kdfopt r:8 -kdfopt p:16 -kdfopt maxmem_bytes:10485760 id-scrypt>
+
```
Comment (levitte) [inline doc/man1/kdf.pod]: I'm sorry, I thoroughly dislike what you've done here.  Not sure what's wrong with the following?

```
=head1 EXAMPLES

Use TLS1-PRF to create a hex-encoded derived key from a secret key and seed:

    openssl kdf -keylen 16 -kdfopt digest:SHA256 -kdfopt key:secret \
                -kdfopt seed:seed TLS1-PRF

Use HKDF to create a hex-encoded derived key from a secret key, salt and info:

    openssl kdf -keylen 10 -kdfopt digest:SHA256 -kdfopt key:secret \
                -kdfopt salt:salt -kdfopt info:label HKDF

Use SSKDF with KMAC to create a hex-encoded derived key from a secret key, salt and info:

    openssl kdf -keylen 64 -kdfopt mac:KMAC128 -kdfopt maclen:20 \
                -kdfopt hexkey:b74a149a161545 -kdfopt hexinfo:348a37a2 \
                -kdfopt hexsalt:3638271ccd68a2 SSKDF

Use SSKDF with HMAC to create a hex-encoded derived key from a secret key, salt and info:

    openssl kdf -keylen 16 -kdfopt mac:HMAC -kdfopt digest:SHA256 \
                -kdfopt hexkey:b74a149a -kdfopt hexinfo:348a37a2 \
                -kdfopt hexsalt:3638271c SSKDF

Use SSKDF with Hash to create a hex-encoded derived key from a secret key, salt and info:

    openssl kdf -keylen 14 -kdfopt digest:SHA256 \
                -kdfopt hexkey:6dbdc23f045488 \
                -kdfopt hexinfo:a1b2c3d4 SSKDF

Use SSHKDF to create a hex-encoded derived key from a secret key, hash and session_id:

    openssl kdf -keylen 16 -kdfopt md:SHA256 \
                -kdfopt hexkey:0102030405 \
                -kdfopt hexxcghash:06090A \
                -kdfopt hexsession_id:01020304 \
                -kdfopt type:A SSHKDF>

Use PBKDF2 to create a hex-encoded derived key from a password and salt:

    openssl kdf -keylen 32 -kdfopt digest:SHA256 -kdfopt pass:password \
                -kdfopt salt:salt -kdfopt iter:2 PBKDF2

Use scrypt to create a hex-encoded derived key from a password and salt:

    openssl kdf -keylen 64 -kdfopt pass:password -kdfopt salt:NaCl \
                -kdfopt N:1024 -kdfopt r:8 -kdfopt p:16 \
                -kdfopt maxmem_bytes:10485760 id-scrypt
```

Comment (slontis) [reply doc/man1/kdf.pod]: The output from perldoc looks like cat vomit.. Will change it :)

File: doc/man1/kdf.pod (hunk 6 above)
Comment (slontis) [reply doc/man1/kdf.pod]: That is the short name (SN) that is associated with the NID for scrypt. Some of the names have id- prepended.

**Comment by slontis** [comment]
> Is this referred to sshkdf specifically? If so where do I need to change things?

static int kdf_sshkdf_ctrl_str(EVP_KDF_IMPL *impl, const char *type, const char *value)
{
    if (value == NULL) {
        KDFerr(KDF_F_KDF_SSHKDF_CTRL_STR, KDF_R_VALUE_MISSING);
        return 0;
    }

    if (strcmp(type, "md") == 0)
        return kdf_md2ctrl(impl, kdf_sshkdf_ctrl, EVP_KDF_CTRL_SET_MD, value);

Change it to:
    if (strcmp(type, "digest") == 0)

And then in the tests it needs to be changed (I think its mainly in evpdigest.txt)

File: test/recipes/20-test_kdf.t (hunk 5 above)
Comment (slontis) [reply test/recipes/20-test_kdf.t]: Added

File: doc/man1/kdf.pod (hunk 18 above)
Comment (levitte) [reply doc/man1/kdf.pod]: Huh? Not sure I understand what you mean with "cat vomit"

Comment (slontis) [reply doc/man1/kdf.pod]: It looks ok with your latest format, it is a pity it doesnt copy/paste very nicely to the commandline for testing purposes.

**Comment by slontis** [comment]
> maybe also add a "see also" to the `-kdf algorithm` section of `pkeyutl(1)`?
I think this should be in a seperate PR if it is done.

> I see your approach with saving all option data first, and only act on them after all arguments have been processed. This is a huge contrast to, for example, `openssl genpkey`, which acts immediately on the options (which means the options must come in a specific order).
> 
> We should probably think about streamlining all apps to behave the same way...

Hmm - well I dont currently process them individually, so the order of the options is still the read in order.

**Comment by simo5** [comment]
> > Is this referred to sshkdf specifically? If so where do I need to change things?
> 
> ```c
> static int kdf_sshkdf_ctrl_str(EVP_KDF_IMPL *impl, const char *type, const char *value)
> {
>     if (value == NULL) {
>         KDFerr(KDF_F_KDF_SSHKDF_CTRL_STR, KDF_R_VALUE_MISSING);
>         return 0;
>     }
> 
>     if (strcmp(type, "md") == 0)
>         return kdf_md2ctrl(impl, kdf_sshkdf_ctrl, EVP_KDF_CTRL_SET_MD, value);
> ```
> 
> Change it to:
> 
> ```c
>     if (strcmp(type, "digest") == 0)
> ```
> 
> And then in the tests it needs to be changed (I think its mainly in evpdigest.txt)
> 
> Also add to test/recipies/20-test_kdf.t (After this PR is merged)

I did leave "md" as alias so that tests do not need to be changed, but I can change them if requested on the new PR.

I'll add a not to change 20-test_kdf.t before merging

File: doc/man1/kdf.pod (hunk 6 above)
Comment (tomato42) [reply doc/man1/kdf.pod]: is there a way to query openssl for the full list of those IDs?

**Comment by tomato42** [comment]
> > maybe also add a "see also" to the `-kdf algorithm` section of `pkeyutl(1)`?
> 
> I think this should be in a seperate PR if it is done.

file #8784

**Comment by slontis** [comment]
ping

**Comment by levitte** [comment]
While looking at this, I realised this warrants an entry in NEWS and in CHANGES.  My approval is withdrawn until they appear.

**Comment by slontis** [comment]
> While looking at this, I realised this warrants an entry in NEWS and in CHANGES. My approval is withdrawn until they appear.

Updated those docs (added mac application entries also)

**Comment by levitte** [comment]
Sorry to nag, but there's a need for a rebase...

**Comment by levitte** [comment]
Merged.

07822c515574223c2958fcddd52191cfbf9aefcd added openssl app 'kdf' and 'mac' to the NEWS and CHANGES docs
c54492ecf8331fd87f6ac163f9fc5b576af128e8 Added app for EVP_KDF

**Pull request closed without merging**